### PR TITLE
Detective video bug revamp

### PR DESCRIPTION
**Pull request by CydiaLamiales**
## About The Pull Request
This revamps the Detective's video bugs.
The detective now starts out with 5 of them instead of 1, they can be planted in more places and the Detective can choose which one to view.
The bugs now hide under floor tiles and can be directly attached to (almost) any piece of clothing.
You can change the bug's ID by using it inhand.
The bugs can also be attached to another carbon's clothes by clicking on them with the bug while aiming the desired bodypart, but this takes a 1 second progress bar. Note that clothes take priority over other bodyparts if they cover them.
As an example, clicking on a hand would prioritize :
Suit that hides gloves -> gloves -> suit that covers arms -> uniform that covers arms.
Any clothing item can have only one bug attached to it, and will have a randomly placed pixel that "pulses" every 60 seconds with a green light.
To remove the bug from a clothing item you must click within 2 pixels of that green light and wait for a 4 second progress bar.
EMPs will disable the video bug for at least a minute, longer scaling with strength, and destroying them with enough severity.
The cryo pod will now store the spyglasses and any tracking bugs, but will remove them from a clothing item and store them separately.

## Why It's Good For The Game
As it is right now, the "pocket protector" video bug that the detective has only works on new players. Its limited applicability and appearance of a rarely used item make it stand out like a sore thumb.
This PR revamps the bugs to me a lot more useful and stealthy, but discoverable by those who pay close attention.
This also greatly elevates the Detective's spying and mystery solving capabilities.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>
Note that the "pulsing" here was shortened to 3 seconda for the sake of showcase.

Applications of the bugs

https://github.com/BeeStation/BeeStation-Hornet/assets/110184118/cbb80b21-0266-40c1-8042-eb0b52b2c24c

Sticking bugs onto different bodypart covering clothes

https://github.com/BeeStation/BeeStation-Hornet/assets/110184118/2d7d7bfc-84dd-44a1-99c3-75ad1b1f106e

Bugs pulsing

https://github.com/BeeStation/BeeStation-Hornet/assets/110184118/5ed66d25-3230-4ec3-b2b4-1b48f8ac404e

Removing a bug by clicking close to the blinking light

https://github.com/BeeStation/BeeStation-Hornet/assets/110184118/3cca93b2-7987-4dff-b56e-8bce6f04763c

</details>

## Changelog
:cl:
add: Added new things
del: Removed the old pocket protector video bug
add: Added the new revamped video bugs for the Detective
balance: Made clothing with attached video bugs now pulse every 60 seconds with a green light
tweak: Made the process of removing video bugs require the user to click on the position of the blinking green light on the clothing item
tweak: Allowed the video bugs to be planted on a worn clothing item by clicking on a mob and aiming at a respective bodypart. Planting on a mob has a 1 second delay.
tweak: Detective's spyglasses and video bugs are now stored by the cryo console
balance: Detective gets 5 video bugs instead of 1
balance: Detective's video bugs now can be disabled by EMPs, and even destroyed by them
balance: Detective's video bugs and spyglasses now can get jammed by uplink radio jammers
tweak: Made the detective's video bugs be hidden under floor tiles
imageadd: added a new video bug icon + overlays
code: made helmets and hardhats call the parent update_appearance proc
code: added a buggable variable which determines if a video bug can be attached to this piece of clothing
/:cl:

### CODE REVIEW

**Comment by DarnTheMarn** [comment]
Pretty sweet, detective getting more gadgets is always cool. The only concern I have is using them roundstart on objectives like handtele, but that's a player issue and it still wouldn't be too terrible.

**Comment by CydiaLamiales** [comment]
> The only concern I have is using them roundstart on objectives like handtele

Can't do that since it's not a clothing item. I guess the CE magboots would be the only ones that'd fall under this category, but you're right about this being a player issue.
I should probably add logging to this so it's easier to keep track of in investigations.

**Comment by Rukofamicom** [comment]
If I'm not mistaken this change in behavior means I can't bug potted plants with a camera anymore.  I quite liked doing this because players rarely check inside of potted plants.

**Comment by Tyranicranger4** [comment]
I love this idea. However, I really think they should have a short do_after to place them. Something long enough to prevent you from driveby bugging people; but short enough to where you can quickly use it on someone standing still.
I think it'd be better both for balancing and it would also be more interesting if detectives were encouraged to distract people so they could bug them.

File: code/__HELPERS/icons.dm [hunk 1]
```diff
@@ -1459,3 +1459,85 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
...
+	var/x_coord = 1
+	var/y_coord = 1
+	var/height = tempicon.Height()
+	var/width = tempicon.Width()
+	var/left_border = width
+	var/right_border = 1
+
+	tempicon.Scale(width, 1) //Flatten the icon into a 1 pixel high line
+	height = tempicon.Height()
+	while(y_coord <= height)
```
Comment (Penwin0) [inline code/__HELPERS/icons.dm]: Does it actually flatten the icon? I might be missunderstanding the proc, but reading the documentation I'm skeptical

Comment (Penwin0) [inline code/__HELPERS/icons.dm]: If it got squished into 1 pixel, we already know the height, which is one, so this should be unecessary

File: code/__HELPERS/icons.dm [hunk 2]
```diff
@@ -1459,3 +1459,85 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
...
+	height = tempicon.Height()
+	width = tempicon.Width()
+	var/bottom_border = height
+	var/top_border = 1
+	x_coord = 1
+	y_coord = 1
+	tempicon.Scale(1, height) //Squash the icon into a 1 pixel wide pillar
+	width = tempicon.Width()
+
+	while(x_coord <= width)
```
Comment (Penwin0) [inline code/__HELPERS/icons.dm]: What I mentioned earlier applies here as well

**Comment by PowerfulBacon** [comment]
Camera and viewer should both be disrupted by signal jammers

File: code/__HELPERS/icons.dm [hunk 3]
```diff
@@ -1458,3 +1458,83 @@ GLOBAL_LIST_EMPTY(friendly_animal_types)
 	animate(src, pixel_x = pixel_x + shiftx, pixel_y = pixel_y + shifty, time = 0.2, loop = duration)
 	pixel_x = initialpixelx
 	pixel_y = initialpixely
+
+/**
+ * Returns a list that contains the leftmost, rightmost, bottom and top pixels in an atom's icon
+ */
+/atom/proc/get_bounding_box()
```
Comment (PowerfulBacon) [inline code/__HELPERS/icons.dm:1465]: This proc need replacement due to image proc performance. I am fine with instead just doing something dodgy and risking it falling off the edge of the icon. You could make the bug only placable on clothing if you wanted to reduce the risk of it being placed on a crowbar or any other thin icons. A solution needs to be found that doesn't use any byond icon procs

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/spyglasses.dm [hunk 4]
```diff
@@ -1,64 +1,305 @@
...
 	var/datum/movement_detector/tracker
+	var/atom/movable/screen/background/cam_background
+
+	var/disabled = FALSE
+
+/obj/item/video_bug/Initialize(mapload)
+	update_overlays()
+	var/tile_overlay = image(icon = icon, icon_state = icon_state_clothing, pixel_x = rand(-12, 12), pixel_y = rand(-12, 12))
+	AddElement(/datum/element/undertile, TRAIT_T_RAY_VISIBLE, INVISIBILITY_OBSERVER, tile_overlay, use_anchor = TRUE)
+	bug_id = "[pick(GLOB.posibrain_names)]-[rand(1,999)]"
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/spyglasses.dm:147]: Posibrain names? would it be better to use the phonetic alphabet (alpha, beta, charlie, delta, etc.)

File: code/game/objects/items/devices/spyglasses.dm [hunk 5]
```diff
@@ -1,64 +1,305 @@
...
+	var/left_border = 1
+	var/right_border = 32
+	var/bottom_border = 1
+	var/top_border = 32
+	var/list/temp_list = target.get_bounding_box()
+	left_border = temp_list["left"]
+	right_border = temp_list["right"]
+	bottom_border = temp_list["bottom"]
+	top_border = temp_list["top"]
+	var/icon/target_icon = getFlatIcon(target)
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/spyglasses.dm:206]: getFlatIcon cannot be used, in theory the get pixels should be cold enough that 10 of them are fine here but you need to use the regular icon and not flatten it as this proc is horrificly slow to the point where 1 proc will ruin this

File: code/game/objects/items/devices/spyglasses.dm [hunk 6]
```diff
@@ -1,64 +1,305 @@
...
+		selected_x = rand(left_border + 2, right_border - 2)
+		selected_y = rand(bottom_border + 2, top_border - 2)
+		if(isnull(target_icon.GetPixel(selected_x, selected_y)))
+			attempts ++
+		else
+			selected_coordinates = TRUE
+			break
+	if(!selected_coordinates)
+		selected_x = 15
+		selected_y = 15
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/spyglasses.dm:221]: Remove the bouding box and assume a bounding box of something like +/-8 pixels in the X and Y. If you can't find one, placing it in the center should always be fine

File: code/game/objects/items/devices/spyglasses.dm [hunk 7]
```diff
@@ -1,64 +1,305 @@
...
+		else
+			return ..()
+	if(iscarbon(A) && !succeeded)
+		if(attach_to_carbon(user, A))
+			succeeded = TRUE
+	if(!succeeded)
+		return ..()
+
+/obj/item/video_bug/emp_act(severity)
+	if(severity > 4)
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/spyglasses.dm:276]: Severity is either 1 or 2

File: code/game/objects/items/devices/spyglasses.dm [hunk 8]
```diff
@@ -1,64 +1,305 @@
...
+		Destroy(src)
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/spyglasses.dm:277]: You can't manually call Destroy(), you have to call qdel

File: code/game/objects/items/devices/spyglasses.dm [hunk 9]
```diff
@@ -1,64 +1,305 @@
...
+		return ..()
+	disabled = TRUE
+	update_appearance(updates = UPDATE_ICON|UPDATE_OVERLAYS)
+	if(attached_to)
+		attached_to.update_appearance(updates = UPDATE_ICON|UPDATE_OVERLAYS)
+	update_view()
+	addtimer(CALLBACK(src, PROC_REF(recover_from_emp)), 60 SECONDS * severity)
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/spyglasses.dm:284]: You need to verify this but I think that 1 severity is high severity and 2 is low severity.

File: code/game/objects/items/devices/spyglasses.dm [hunk 10]
```diff
@@ -1,64 +1,305 @@
...
+/obj/item/video_bug/pre_attack(atom/A, mob/living/user, params)
+	var/succeeded = FALSE
+	if(istype(A, /obj/item/clothing))
+		if(attach_to_clothing_item(user, A, TRUE))
+			to_chat(user, "<span class='notice'>You stealthily plant \the [src] on \the [A]!</span>")
+			succeeded = TRUE
+		else
+			return ..()
+	if(iscarbon(A) && !succeeded)
+		if(attach_to_carbon(user, A))
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/spyglasses.dm:270]: This definitely needs a do_after action timer

File: code/modules/clothing/clothing.dm [hunk 11]
```diff
@@ -453,6 +474,40 @@ BLIND     // can't see anything
 	if(!lavaland_equipment_pressure_check(T))
 		. *= high_pressure_multiplier
 
+/obj/item/clothing/emp_act(severity)
+	if(tracking_bug)
+		tracking_bug.emp_act(severity)
+	. = ..()
```
Comment (PowerfulBacon) [inline code/modules/clothing/clothing.dm:480]: Register a signal instead on the tracking bug when you attach it to some clothing

File: code/modules/clothing/clothing.dm [hunk 12]
```diff
@@ -453,6 +474,40 @@ BLIND     // can't see anything
...
+		if(found_x >= 0 && found_x <= 2 && found_y >= 0 && found_y <= 2)
+			to_chat(user, "<span class='notice'>Your fingers find something stuck to \the [src], and you try to rip it away!</span>")
+			if(!(src in user.do_afters))
+				if(do_after(user, 4 SECONDS, src))
+					to_chat(user, "<span class='notice'>You dislodge \the [tracking_bug] that was planted on \the [src]!</span>")
+					tracking_bug.forceMove(drop_location())
+					user.put_in_hands(tracking_bug)
+					tracking_bug.detach_from_clothing(user, src)
+			return COMPONENT_NO_ATTACK
+	return ..()
```
Comment (PowerfulBacon) [inline code/modules/clothing/clothing.dm:509]: Isn't this handled by pre_attaack of the video bug? Overriding click is horrible

File: code/game/machinery/cryopod.dm [hunk 13]
```diff
@@ -308,6 +310,17 @@ GLOBAL_LIST_EMPTY(cryopod_computers)
...
+		if(istype(W, /obj/item/clothing))
+			var/obj/item/clothing/cloth = W
+			if(cloth.tracking_bug)
+				cloth.tracking_bug.forceMove(drop_location())
+				if(control_computer && control_computer.allow_items)
+					control_computer.frozen_items += W
+					mob_occupant.transferItemToLoc(cloth.tracking_bug, control_computer, TRUE)
+				else
+					mob_occupant.transferItemToLoc(cloth.tracking_bug, loc, TRUE)
+				cloth.tracking_bug.detach_from_clothing(mob_occupant, cloth)
```
Comment (PowerfulBacon) [inline code/game/machinery/cryopod.dm:322]: This seems way too specific, is there a better way of doing this that doesnt require specificly type checking for the bug when they cryo?

File: code/game/objects/items/devices/spyglasses.dm [hunk 14]
```diff
@@ -1,64 +1,305 @@
...
+		if(isnull(target_icon.GetPixel(selected_x, selected_y)))
+			attempts ++
+		else
+			selected_coordinates = TRUE
+			break
+	if(!selected_coordinates)
+		selected_x = 15
+		selected_y = 15
+	user.temporarilyRemoveItemFromInventory(src)
+	forceMove(null)
```
Comment (PowerfulBacon) [inline code/game/objects/items/devices/spyglasses.dm:223]: This feels like it is going to cause loads of issues, like with the cryo pod. You can probably make stuff like the cryo-pod handled better by making this a component instead and just qdeling the bug and referencing the component instead.
A majority of this code is written in a way that wants to be a component but is suffering from not being a component and is forced to insert specific type checks all over the codebase to componensate. This code needs to be modular and a component would help with that.

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]

File: code/game/objects/items/devices/spyglasses.dm (hunk 4 above)
Comment (CydiaLamiales) [reply code/game/objects/items/devices/spyglasses.dm:147]: Phonetic alphabet is already used for network ID names.

**Comment by CydiaLamiales** [comment]
This pretty much requires a whole rewrite of the code so I'll shelve it for now

**Pull request closed without merging**